docker/daemon/daemon_dest.go

```diff
@@ -157,15 +157,26 @@ func (d *daemonImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobI
 		defer os.Remove(streamCopy.Name())
 		defer streamCopy.Close()
 
-		size, err := io.Copy(streamCopy, stream)
+		// If we need to compute a digest.
+		digester := digest.Canonical.New()
+		tee := io.TeeReader(stream, digester.Hash())
+
+		size, err := io.Copy(streamCopy, tee)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
 		_, err = streamCopy.Seek(0, os.SEEK_SET)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+
+		if inputInfo.Digest == "" {
+			// In some (hacky) cases a caller might not know the size, but will
+			// know the digest (where "know" == "have a fake digest").
```

Comment:
Thanks for splitting this. Smaller PRs have less to block them from getting merged :)

(Really the point is that the API as defined expect the digest to either be unknown or valid, and the rest of this PR should make that true again. If we need change the API for other reasons, sure, let’s talk about that, but _actually explicitly talk about  it_ instead of adding a weird corner case which is unjustified by the rest of the code in the repo.)